test/run-tests

```diff
@@ -853,31 +885,104 @@ def main():
     parser.add_argument(
         "--use-images",
         default="*",
-        help="Test pull request only against images matching this pattern",
+        help=(
+            "Test pull request only against images matching these patterns.  "
+            "This is a comma delimited list of fnmatch patterns which will be "
+            "applied to each image name and source.  If any of the patterns "
+            "match the name or source, the image will be included in testing."
+        ),
     )
     parser.add_argument(
         "--dry-run",
         help="Do not update pull request status or upload artifacts",
         default=False,
         action="store_true",
     )
+    parser.add_argument(
+        "--keep-results",
+        help="For debugging - do not remove the temp results directory",
+        default=False,
+        action="store_true",
+    )
 
     args = parser.parse_args()
 
     # default values for config
     config = {"repositories": [], "images": [], "name": "linux-system-roles-test"}
 
+    # check for settings via env. vars
+    args.secrets = os.environ.get("TEST_HARNESS_SECRETS", args.secrets)
```

Comment:
Sorry, did not spot this before now. It seems that config variables have the highest precendence, here. For me it seems to be more commont to have the following precendence from lowest to highest:

1) internal defaults
2) environment variables
3) config file
4) command-line arguments

not 100% sure about 2 vs 3 since it might not be that common to have that many ways to specify the same settings but since the variables are least visible, they should probably have the least precedence for user-supplied settings.

Reply:
I'm not sure what you mean by "config variables have the highest precedence, here.".  Are "config variables" the settings from the config.json file?  If so, there aren't any command-line arguments which allow setting those.  The only way to override those are with the `TEST_HARNESS_CONFIG_VARIANT` and `TEST_HARNESS_REPOSITORIES` env. vars.

The way it is coded now is that environment variables have the highest precedence.  For example:
```python
    args.secrets = os.environ.get("TEST_HARNESS_SECRETS", args.secrets)
```
This means - the value of `args.secrets` will be taken from the environment variable `TEST_HARNESS_SECRETS` if it is set, otherwise, use the value from the `--secrets` command-line argument if set, otherwise, use the default value.

So the precedence is:
* environment variables - highest - will override everything
* command-line arguments - can be overridden by environment variables
* config file - can be overridden by command-line arguments and environment variables
* internal defaults

This is better for the use case of running in a pod in kubernetes/openshift - it is much more difficult to change the command line arguments than to change the environment variables in a deployment.  It is a common idiom in kubernetes/openshift to allow customization via environment variables - setting an environment variable in a deployment will trigger a redeploy.

Reply:
Sorry, I meant "environment variables" when I wrote "config variables" which is surprising to me, I guess code like this would be less surprising:

```python
parser.add_argument("--secrets", default=os.environ.get("TEST_HARNESS_SECRETS", "/secrets"), help="Directory with secrets")
```

> This is better for the use case of running in a pod in kubernetes/openshift - it is much more difficult to change the command line arguments than to change the environment variables in a deployment. It is a common idiom in kubernetes/openshift to allow customization via environment variables - setting an environment variable in a deployment will trigger a redeploy.

I see. But this does not require the environment variables to take precedence over command-line arguments since I guess it would be typical to deploy it without any arguments and configure it completely with variables. As far as I understand, Ansible is also allowing to use `--inventory` to override `ANSIBLE_INVENTORY` which than override the config file. So I guess I would actually find this least surprising:

* command-line arguments - will override everything:
* environment variables - highest - will override config file and defaults
* config file - can be overridden by command-line arguments and environment variables
* internal defaults

There is also a python pattern to first parse only the command-line flag to find the config file and then use this to get the defaults for a second parser run to check for remaining arguments: https://stackoverflow.com/questions/3609852/which-is-the-best-way-to-allow-configuration-options-be-overridden-at-the-comman

Reply:
> There is also a python pattern to first parse only the command-line flag to find the config file and then use this to get the defaults for a second parser run to check for remaining arguments

I didn't implement that, but I think this https://github.com/linux-system-roles/test-harness/pull/87/commits/fc418af4d8a50db913a4ab98b2b8f0500af98a9e makes the precedence as you suggested, and makes the usage of command line arguments and environment variables more consistent.